File: backend/src/services/ollama.service.ts
import { env } from '../config/env.js';
import {
  approxTokens,
  withLogging,
  type LlmLogContext,
} from './observability/llm-logger.js';

export interface OllamaMessage {
  role: 'system' | 'user' | 'assistant' | 'tool';
  content: string;
}

interface OllamaStreamChunk {
  model: string;
  message?: { role: string; content: string };
  response?: string;
  done: boolean;
  prompt_eval_count?: number;
  eval_count?: number;
}

interface OllamaChatResponse {
  model: string;
  message: { role: string; content: string };
  done: boolean;
  prompt_eval_count?: number;
  eval_count?: number;
}

export interface ChatOptions {
  onChunk?: (text: string) => void;
  signal?: AbortSignal;
  temperature?: number;
  topP?: number;
  numCtx?: number;
  /** Ollama's `format` — either the literal string "json" or a JSON schema object. */
  format?: 'json' | Record<string, unknown>;
  /** Route label for observability; defaults to 'chat'. */
  route?: string;
  logContext?: LlmLogContext;
}

/**
 * Primary chat entry point. Streams when `onChunk` is set, otherwise returns the
 * full completion in one shot. Always goes through the LLM logger so every call
 * lands in LlmCall + Prometheus.
 */
export async function chatCompletion(
  messages: OllamaMessage[],
  onChunkOrOptions?: ((text: string) => void) | ChatOptions,
  legacyOptions?: ChatOptions
): Promise<string> {
  const options: ChatOptions =
    typeof onChunkOrOptions === 'function'
      ? { onChunk: onChunkOrOptions, ...(legacyOptions ?? {}) }
      : (onChunkOrOptions ?? {});

  const route = options.route ?? 'chat';

  return withLogging<string>(
    route,
    async () => {
      const body: Record<string, unknown> = {
        model: env.OLLAMA_MODEL,
        messages,
        stream: !!options.onChunk,
      };
      const ollamaOptions: Record<string, unknown> = {};
      if (options.temperature !== undefined) ollamaOptions.temperature = options.temperature;
      if (options.topP !== undefined) ollamaOptions.top_p = options.topP;
      if (options.numCtx !== undefined) ollamaOptions.num_ctx = options.numCtx;
      if (Object.keys(ollamaOptions).length) body.options = ollamaOptions;
      if (options.format) body.format = options.format;

      const response = await fetch(`${env.OLLAMA_BASE_URL}/api/chat`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
        signal: options.signal,
      });

      if (!response.ok) {
        throw new Error(`Ollama API error: ${response.status} ${response.statusText}`);
      }

      let full = '';
      let promptTokens: number | undefined;
      let completionTokens: number | undefined;

      if (options.onChunk && response.body) {
        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        while (true) {
          const { done, value } = await reader.read();
          if (done) break;
          const lines = decoder.decode(value, { stream: true }).split('\n').filter(Boolean);
          for (const line of lines) {
            try {
              const chunk: OllamaStreamChunk = JSON.parse(line);
              if (chunk.message?.content) {
                full += chunk.message.content;
                options.onChunk(chunk.message.content);
              }
              if (chunk.done) {
                promptTokens = chunk.prompt_eval_count;
                completionTokens = chunk.eval_count;
              }
            } catch {
              // Ollama occasionally splits JSON across chunks — skip garbage.
            }
          }
        }
      } else {
        const data = (await response.json()) as OllamaChatResponse;
        full = data.message?.content ?? '';
        promptTokens = data.prompt_eval_count;
        completionTokens = data.eval_count;
      }

      return {
        value: full,
        completion: full,
        prompt: { model: env.OLLAMA_MODEL, messages, options: ollamaOptions, format: options.format ?? null },
        promptTokens: promptTokens ?? approxTokens(messages.map((m) => m.content).join('\n')),
        completionTokens: completionTokens ?? approxTokens(full),
      };
    },
    { ...(options.logContext ?? {}), model: env.OLLAMA_MODEL }
  );
}

/**
 * Structured chat completion: forces Ollama to emit JSON matching `schema`.
 * Throws if the response cannot be JSON-parsed — callers should surround in
 * their own retry/fallback. qwen2.5:7b-instruct complies well with this.
 */
export async function chatCompletionStructured<T = unknown>(
  messages: OllamaMessage[],
  schema: Record<string, unknown>,
  options: Omit<ChatOptions, 'format' | 'onChunk'> = {}
): Promise<T> {
  const raw = await chatCompletion(messages, {
    ...options,
    format: schema,
    route: options.route ?? 'chat.structured',
  });
  const trimmed = raw.trim();
  try {
    return JSON.parse(trimmed) as T;
  } catch {
    const firstBrace = trimmed.indexOf('{');
    const lastBrace = trimmed.lastIndexOf('}');
    if (firstBrace !== -1 && lastBrace > firstBrace) {
      return JSON.parse(trimmed.slice(firstBrace, lastBrace + 1)) as T;
    }
    throw new Error(`Structured output is not valid JSON: ${trimmed.slice(0, 200)}`);
  }
}

/**
 * Embed one or more texts via Ollama's /api/embed endpoint.
 * nomic-embed-text returns 768-dim vectors. Returns a flat array of vectors
 * aligned with input order.
 */
export async function embed(
  input: string | string[],
  options: { signal?: AbortSignal; logContext?: LlmLogContext } = {}
): Promise<number[][]> {
  const texts = Array.isArray(input) ? input : [input];
  if (texts.length === 0) return [];

  return withLogging<number[][]>(
    'embed',
    async () => {
      const response = await fetch(`${env.OLLAMA_BASE_URL}/api/embed`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          model: env.OLLAMA_EMBEDDING_MODEL,
          input: texts,
        }),
        signal: options.signal,
      });

      if (!response.ok) {
        throw new Error(`Ollama embed error: ${response.status} ${response.statusText}`);
      }

      const data = (await response.json()) as {
        embeddings: number[][];
        prompt_eval_count?: number;
      };

      if (!Array.isArray(data.embeddings) || data.embeddings.length !== texts.length) {
        throw new Error(`Ollama embed returned ${data.embeddings?.length ?? 0} vectors for ${texts.length} inputs`);
      }

      return {
        value: data.embeddings,
        completion: '',
        prompt: { model: env.OLLAMA_EMBEDDING_MODEL, inputCount: texts.length },
        promptTokens: data.prompt_eval_count ?? approxTokens(texts.join('\n')),
        completionTokens: 0,
      };
    },
    { ...(options.logContext ?? {}), model: env.OLLAMA_EMBEDDING_MODEL }
  );
}

export function buildSystemPrompt(courseName?: string, topicTitle?: string): string {
  let context = '';
  if (courseName) context += `Course: "${courseName}". `;
  if (topicTitle) context += `Topic: "${topicTitle}". `;

  return `You are an intelligent academic tutor for university students on the Cognitive Copilot platform.
${context ? `Context: ${context}` : ''}
You help students understand topics deeply, generate quizzes, explain concepts clearly,
suggest study materials, and adapt to the student's level.
Be concise, structured, and pedagogically sound.
When generating quizzes, output strict JSON: { "questions": [{"id": "q1", "question": "...", "options": ["A","B","C","D"], "correct": "A"}] }
When not generating quizzes, use markdown formatting for clarity.`;
}

File: backend-py/app/services/ollama_service.py
from __future__ import annotations

import asyncio
import json
import logging
import time
from typing import Any, AsyncGenerator, Callable, Optional

import httpx

from app.core.config import settings
from app.models.enums import LlmCallStatus

logger = logging.getLogger(__name__)

OllamaMessage = dict  # {"role": str, "content": str}


def approx_tokens(text: str) -> int:
    return max(1, len(text) // 4)


def build_system_prompt(course_name: str | None = None, topic_title: str | None = None) -> str:
    context = ""
    if course_name:
        context += f'Course: "{course_name}". '
    if topic_title:
        context += f'Topic: "{topic_title}". '
    return (
        "You are an intelligent academic tutor for university students on the Cognitive Copilot platform.\n"
        + (f"Context: {context}\n" if context else "")
        + "You help students understand topics deeply, generate quizzes, explain concepts clearly,\n"
        "suggest study materials, and adapt to the student's level.\n"
        "Be concise, structured, and pedagogically sound.\n"
        'When generating quizzes, output strict JSON: { "questions": [{"id": "q1", "question": "...", "options": ["A","B","C","D"], "correct": "A"}] }\n'
        "When not generating quizzes, use markdown formatting for clarity."
    )


async def _log_llm_call(
    route: str,
    model: str,
    prompt_payload: dict,
    completion: str,
    prompt_tokens: int,
    completion_tokens: int,
    latency_ms: int,
    status: LlmCallStatus,
    error_msg: str | None,
    user_id: str | None,
) -> None:
    from app.db.session import AsyncSessionLocal
    from app.models.misc import LlmCall
    try:
        async with AsyncSessionLocal() as session:
            call = LlmCall(
                user_id=user_id,
                route=route,
                model=model,
                prompt=prompt_payload,
                completion=completion,
                prompt_tokens=prompt_tokens,
                completion_tokens=completion_tokens,
                latency_ms=latency_ms,
                status=status,
                error_msg=error_msg,
            )
            session.add(call)
            await session.commit()
    except Exception as exc:
        logger.warning("LLM call logging failed: %s", exc)


async def chat_completion(
    messages: list[OllamaMessage],
    route: str = "chat",
    temperature: float | None = None,
    top_p: float | None = None,
    num_ctx: int | None = None,
    format: Any = None,
    user_id: str | None = None,
    on_chunk: Callable[[str], None] | None = None,
    signal: asyncio.Event | None = None,
) -> str:
    body: dict = {"model": settings.OLLAMA_MODEL, "messages": messages, "stream": on_chunk is not None}
    ollama_opts: dict = {}
    if temperature is not None:
        ollama_opts["temperature"] = temperature
    if top_p is not None:
        ollama_opts["top_p"] = top_p
    if num_ctx is not None:
        ollama_opts["num_ctx"] = num_ctx
    if ollama_opts:
        body["options"] = ollama_opts
    if format is not None:
        body["format"] = format

    t0 = time.monotonic()
    full = ""
    prompt_tokens = 0
    completion_tokens = 0
    status = LlmCallStatus.OK
    error_msg: str | None = None

    try:
        if on_chunk:
            async with httpx.AsyncClient(timeout=httpx.Timeout(300.0, connect=30.0)) as client:
                async with client.stream("POST", f"{settings.OLLAMA_BASE_URL}/api/chat", json=body) as resp:
                    resp.raise_for_status()
                    async for line in resp.aiter_lines():
                        if not line:
                            continue
                        try:
                            chunk = json.loads(line)
                            content = (chunk.get("message") or {}).get("content", "")
                            if content:
                                full += content
                                on_chunk(content)
                            if chunk.get("done"):
                                prompt_tokens = chunk.get("prompt_eval_count", 0) or 0
                                completion_tokens = chunk.get("eval_count", 0) or 0
                        except json.JSONDecodeError:
                            pass
        else:
            async with httpx.AsyncClient(timeout=httpx.Timeout(300.0, connect=30.0)) as client:
                r = await client.post(f"{settings.OLLAMA_BASE_URL}/api/chat", json=body)
                r.raise_for_status()
                data = r.json()
                full = (data.get("message") or {}).get("content", "")
                prompt_tokens = data.get("prompt_eval_count", 0) or 0
                completion_tokens = data.get("eval_count", 0) or 0

    except Exception as exc:
        status = LlmCallStatus.ERROR
        error_msg = str(exc)
        _rethrow_with_friendly_msg(exc)

    finally:
        latency_ms = round((time.monotonic() - t0) * 1000)
        if not prompt_tokens:
            prompt_tokens = approx_tokens("\n".join(m.get("content", "") for m in messages))
        if not completion_tokens:
            completion_tokens = approx_tokens(full)
        asyncio.create_task(_log_llm_call(
            route=route,
            model=settings.OLLAMA_MODEL,
            prompt_payload={"model": settings.OLLAMA_MODEL, "messages": messages},
            completion=full,
            prompt_tokens=prompt_tokens,
            completion_tokens=completion_tokens,
            latency_ms=latency_ms,
            status=status,
            error_msg=error_msg,
            user_id=user_id,
        ))

    return full


async def chat_completion_structured(
    messages: list[OllamaMessage],
    schema: dict,
    route: str = "chat.structured",
    temperature: float | None = None,
    user_id: str | None = None,
) -> Any:
    raw = await chat_completion(messages, route=route, temperature=temperature, format=schema, user_id=user_id)
    trimmed = raw.strip()
    try:
        return json.loads(trimmed)
    except json.JSONDecodeError:
        first = trimmed.find("{")
        last = trimmed.rfind("}")
        if first != -1 and last > first:
            return json.loads(trimmed[first:last + 1])
        raise ValueError(f"Structured output is not valid JSON: {trimmed[:200]}")


def fit_embedding_to_db_dim(vec: list[float], target_dim: int | None = None) -> list[float]:
    """Pad/truncate to match pgvector column (Ollama model output may differ from DB width)."""
    if target_dim is None:
        target_dim = settings.OLLAMA_EMBEDDING_DIM
    if target_dim <= 0 or not vec:
        return vec
    n = len(vec)
    if n == target_dim:
        return vec
    if n > target_dim:
        if n - target_dim > 8:
            logger.warning("Truncating query/document embedding from %d to %d dimensions", n, target_dim)
        return vec[:target_dim]
    return vec + [0.0] * (target_dim - n)


def _parse_ollama_embed_response(data: dict) -> list[list[float]]:
    """Normalize /api/embed JSON across Ollama versions."""
    raw = data.get("embeddings")
    if raw is not None and isinstance(raw, list) and len(raw) > 0:
        return raw
    one = data.get("embedding")
    if one is not None and isinstance(one, list) and len(one) > 0:
        if isinstance(one[0], (int, float)):
            return [list(one)]
        if isinstance(one[0], list):
            return list(one)
    raise ValueError(f"Ollama /api/embed returned no vectors (keys={list(data.keys())})")


_EMBED_BATCH_SIZE = 24


async def _embed_batch(texts: list[str], user_id: str | None) -> list[list[float]]:
    """Single /api/embed call; kept small so Ollama does not time out on long documents."""
    if not texts:
        return []
    t0 = time.monotonic()
    status = LlmCallStatus.OK
    error_msg: str | None = None
    embeddings: list[list[float]] = []

    try:
        async with httpx.AsyncClient(timeout=httpx.Timeout(120.0, connect=30.0)) as client:
            r = await client.post(
                f"{settings.OLLAMA_BASE_URL}/api/embed",
                json={"model": settings.OLLAMA_EMBEDDING_MODEL, "input": texts},
            )
            r.raise_for_status()
            embeddings = _parse_ollama_embed_response(r.json())
        if len(embeddings) != len(texts):
            raise ValueError(
                f"Ollama embed count mismatch: got {len(embeddings)} vectors for {len(texts)} inputs"
            )
        dim = settings.OLLAMA_EMBEDDING_DIM
        embeddings = [fit_embedding_to_db_dim(v, dim) for v in embeddings]
    except Exception as exc:
        status = LlmCallStatus.ERROR
        error_msg = str(exc)
        raise
    finally:
        latency_ms = round((time.monotonic() - t0) * 1000)
        prompt_tokens = approx_tokens("\n".join(texts))
        asyncio.create_task(_log_llm_call(
            route="embed",
            model=settings.OLLAMA_EMBEDDING_MODEL,
            prompt_payload={"model": settings.OLLAMA_EMBEDDING_MODEL, "inputCount": len(texts)},
            completion="",
            prompt_tokens=prompt_tokens,
            completion_tokens=0,
            latency_ms=latency_ms,
            status=status,
            error_msg=error_msg,
            user_id=user_id,
        ))

    return embeddings


async def embed(input: str | list[str], user_id: str | None = None) -> list[list[float]]:
    texts = [input] if isinstance(input, str) else input
    if not texts:
        return []
    out: list[list[float]] = []
    for i in range(0, len(texts), _EMBED_BATCH_SIZE):
        batch = texts[i : i + _EMBED_BATCH_SIZE]
        out.extend(await _embed_batch(batch, user_id))
    return out


async def chat_completion_stream(
    messages: list[OllamaMessage],
    route: str = "chat",
    temperature: float | None = None,
    user_id: str | None = None,
) -> AsyncGenerator[str, None]:
    body: dict = {
        "model": settings.OLLAMA_MODEL,
        "messages": messages,
        "stream": True,
    }
    if temperature is not None:
        body["options"] = {"temperature": temperature}

    t0 = time.monotonic()
    full = ""
    prompt_tokens = 0
    completion_tokens = 0
    status = LlmCallStatus.OK
    error_msg: str | None = None

    try:
        async with httpx.AsyncClient(timeout=httpx.Timeout(300.0, connect=30.0)) as client:
            async with client.stream("POST", f"{settings.OLLAMA_BASE_URL}/api/chat", json=body) as resp:
                resp.raise_for_status()
                async for line in resp.aiter_lines():
                    if not line:
                        continue
                    try:
                        chunk = json.loads(line)
                        content = (chunk.get("message") or {}).get("content", "")
                        if content:
                            full += content
                            yield content
                        if chunk.get("done"):
                            prompt_tokens = chunk.get("prompt_eval_count", 0) or 0
                            completion_tokens = chunk.get("eval_count", 0) or 0
                    except json.JSONDecodeError:
                        pass
    except Exception as exc:
        status = LlmCallStatus.ERROR
        error_msg = str(exc)
        yield f"\n\n⚠️ AI service error: {exc}"
    finally:
        latency_ms = round((time.monotonic() - t0) * 1000)
        if not prompt_tokens:
            prompt_tokens = approx_tokens("\n".join(m.get("content", "") for m in messages))
        if not completion_tokens:
            completion_tokens = approx_tokens(full)
        asyncio.create_task(_log_llm_call(
            route=route,
            model=settings.OLLAMA_MODEL,
            prompt_payload={"model": settings.OLLAMA_MODEL, "messages": messages},
            completion=full,
            prompt_tokens=prompt_tokens,
            completion_tokens=completion_tokens,
            latency_ms=latency_ms,
            status=status,
            error_msg=error_msg,
            user_id=user_id,
        ))
